### PR TITLE
Add end2end tests for tree tabs feature

### PR DESCRIPTION
**Pull request by pylbrecht**
---

Fixes part of #8072

### CODE REVIEW

**Comment by toofar** [comment]
Regarding merging the two tab compare functions, the main thing for me is whether we go with the more nuanced comparsions in the existing method or the string based ones in the new method. I did a quick test:

test: active suffix on wrong line
new:

    E       AssertionError: assert '- http://loc...numbers/2.txt' == '- http://loc....txt (active)'
    E         Skipping 33 identical leading characters in diff, use -v to show
    E         - bers/1.txt
    E         + bers/1.txt (active)
    E         -   - http://localhost:54629/data/numbers/2.txt (active)
    E         ? --                                           ---------
    E         + - http://localhost:54629/data/numbers/2.txt
existing:

    E               KeyError: 'active'

test: tabs in wrong order:
new:

    E       AssertionError: assert '- http://loc...numbers/1.txt' == '- http://loc...numbers/2.txt'
    E         - - http://localhost:60957/data/numbers/1.txt
    E         - - http://localhost:60957/data/numbers/2.txt
    E         + - http://localhost:60957/data/numbers/2.txt
    E         ?                                            +
    E         + - http://localhost:60957/data/numbers/1.txt
existing:

    E           AssertionError: assert 'http://local...numbers/1.txt' == 'http://local...numbers/2.txt'
    E             - http://localhost:56427/data/numbers/2.txt
    E             ?                                     ^
    E             + http://localhost:56427/data/numbers/1.txt
    E             ?                                     ^

From that I draw the conclusions:
1. there is room for improvement in both (displaying a unified git diff would be pretty good I think
2. a string based comparison works great for missing suffixes
3. the line based comparison is a little better for the wrong order case (it has the little arrows)!

So I plan to:
1. merge the two implementations and just compare line by line, no special handling of suffxies
2. try with a python difflib library and see if that's even better.

My tests were simply:

    Scenario: test tree tab error reporting
        When I open data/numbers/1.txt
        And I open data/numbers/2.txt in a new tab
        Then the following tree tabs should be open:
            - data/numbers/2.txt
            - data/numbers/1.txt

    Scenario: test flat tab error reporting
        When I open data/numbers/1.txt
        And I open data/numbers/2.txt in a new tab
        Then the following tabs should be open:
            - data/numbers/2.txt
            - data/numbers/1.tx

**Comment by toofar** [comment]
Regarding the above comment about error output, using difflib (inspired by this comment https://github.com/pytest-dev/pytest/issues/1531#issuecomment-723590313), the error output I get is:

```
test: active suffix on wrong line
E       AssertionError: Session Mismatch
E         --- expected
E         +++ actual
E         @@ -1,2 +1,2 @@
E         - http://localhost:42177/data/numbers/1.txt (active)
E         - http://localhost:42177/data/numbers/2.txt
E         + http://localhost:42177/data/numbers/1.txt
E         + http://localhost:42177/data/numbers/2.txt (active)
E
E       assert False

test: tabs in wrong order:
E       AssertionError: Session Mismatch
E         --- expected
E         +++ actual
E         @@ -1,2 +1,2 @@
E         + http://localhost:42003/data/numbers/1.txt
E           http://localhost:42003/data/numbers/2.txt
E         - http://localhost:42003/data/numbers/1.txt
E
E       assert False
```

<details>
<summary>code</summary>

<pre><code
>    any_diff = any(not l.startswith(" ") for l in diff)
    assert not any_diff, "Session Mismatch\n" + "".join(difflib.unified_diff(
        expected_tabs,
        actual,
        fromfile="expected",
        tofile="actual"
    ))
</code></pre>
</details>

Pros:
* familiar unified diff syntax
* full context
* no question marks in the +/- column

Cons:
* no arrows and underlines of which parts of the line differ

I think having the little indicators telling you where the issue is is a huge win and having the unified diff syntax, despite being familiar, doesn't draw your eye to the bit that needs addressing faster.
It would be nice if we could get pyest to print out the full diff for certain assertions without having to pass `-v` and get much more logs. I tried using `difflib.ndiff()` too, which looks like the same format pytest uses. But is didn't consistently apply the little pointers :cry:

I think for now I'll just remove the hyphens from the start of the lines to make the diff less noisy (like the previous implementation did) and leave it at that. Even though I think showing the full context would be nice.

**Comment by toofar** [comment]
@pylbrecht I cherry picked the "Fix tab widget benchmark tests" commit to the base branch now, so you might want to rebase.

File: tests/end2end/features/conftest.py [hunk 1]
```diff
@@ -620,55 +605,104 @@ def check_contents_json(quteproc, text):
...
+                prefix="  " * indentation,
+                collapsed=tree_node["collapsed"],
+            )
+        else:
+            tree_node = node
+
+        for uid in tree_node["children"]:
+            yield from tree_to_str(tree_data[uid], tree_data, indentation + 1)
+
+    is_tree_tab_window = "treetab_root" in window
```
Comment (pylbrecht) [inline tests/end2end/features/conftest.py:672]: I feel at some point we will be better off encapsulating treetab logic somewhere (I believe we already talked about this to some extent at least). Right now one has to understand treetab internals to ask for certain properties (e.g. if a window is a treetab window).

At this stage of development I am more than ok with domain logic floating around, though. Just leaving my thought here as a reminder for later.

**Comment by toofar** [comment]
Lots more tests to come, but with this foundation we can start adding them on other branches too as we work through other tickets. :rocket:

**Pull request merged**